### DISCUSSION TITLE
Flutter - Firebase (JWT) and system time error

### DISCUSSION THREAD

**Discussion by iNima** [Web3Auth/Web3Auth#757] [Help]
Hi there, I'm trying to use my Firebase JWT token id to create a user and wallet with Web3Auth in Flutter (Custom Authentication).
(By the way, I needed to use `jwt-login-fixes` branch to be able to reach to this point. With the master branch I just got a blank screen with the Web3Auth logo at the bottom of the page without reconstructing the private key message and spinner)
I think I've done everything based on the documents and guides but I received this error when Web3Auth is trying to reconstruct the private-key.

**Tested device:** Samsung Galaxy S20
**Android version:** 12

``There seems to be some bug in the code. Please contact support to fix this.Your system time is incorrect. Please correct your system time and login again"
``
Also sometimes I get this error instead of the above message (with the same code!):
``Unable to detect login share from the Auth Network. This may be due to slow internet connection. Check your internet speed and try again.``

The errors that I got were exactly the same as this one: 
[https://github.com/Web3Auth/web3auth-web/discussions/515#discussioncomment-3624987](url)

Could you please let me know what is the issue?

This is my code in Flutter:
```   
    if (Platform.isAndroid) {
      redirectUrl = Uri.parse('w3a://com.example.w3aflutter/auth');
    } else if (Platform.isIOS) {
      redirectUrl = Uri.parse('com.example.w3aflutter://openlogin');
    } else {
      throw UnKnownException('Unknown platform');
    }

   final loginConfig = HashMap<String, LoginConfigItem>();
    loginConfig['jwt'] = LoginConfigItem(
        verifier: "verifier_name",
        typeOfLogin: TypeOfLogin.jwt,
        name: "Custom JWT Login",
        clientId: "The_web3auth_dashboard_client_id",
    );

    await Web3AuthFlutter.init(Web3AuthOptions(
        clientId: "The_web3auth_dashboard_client_id",
        network: Network.cyan,
        redirectUrl: redirectUrl,
        whiteLabel: WhiteLabelData(
            dark: true, name: "Web3Auth Flutter App", theme: themeMap),
        loginConfig: loginConfig
    ),);

  Future<Web3AuthResponse> _withJWT() {
    return Web3AuthFlutter.login(LoginParams(
        loginProvider: Provider.jwt,
        extraLoginOptions: ExtraLoginOptions(
            verifierIdField: 'sub',
            response_type: 'token',
            id_token: idToken, //Received from firebase login
            domain: "http://localhost:3000")));
  }

**Comment by nisa10880** [comment]
I also have the same issues.

**Comment by himanshuchawla009** [comment]
There can be two reasons of this:-

1. Either your system clock is really out of sync.
2.  The jwt token you are passing in web3auth sdk is older than 1 min. Web3Auth network only accepts unique tokens which are not older than 60 sec so make sure to pass a unique freshly generated jwt token.

If above doesn't solves your issue, plz attach console logs for more debugging on our side.

**Reply by shahbaz17** [reply]
It looks like Option 2 is the cause. While testing, I was also getting the same when an expired token was used.

@iNima Are you using a fresh JWT Token? And Can you share(code-snippet) how you are generating idToken from Firebase?

**Comment by shahbaz17** [comment]
Hey all,

We have released `v1.0.3` that fixes the JWT Login issue. Please update your code to use the latest version.

https://pub.dev/packages/web3auth_flutter

**Reply by nisa10880** [reply]
I used this release and the web view redirects me to torus help chat. I have this error message. "Unable to detect login share from the Auth Network. This may due to slow internet connection. Check your internet speed and try again."
The message is not clear, my internet speed is very good.

**Reply by nisa10880** [reply]
I used phone auth method from firebase to get the id_token.

**Reply by shahbaz17** [reply]
Can you share the code snippet of the same?

**Reply by shahbaz17** [reply]
Please check https://firebase.google.com/docs/reference/js/v8/firebase.User#getidtoken

Are you using `user.getIdToken(false)` in your code? if so, please pass true to get the latest/fresh token each time.

**Reply by nisa10880** [reply]
This is my initialisation function:
```
  Future<void> initPlatformState() async {
    
    HashMap<String, LoginConfigItem>? loginConfig = HashMap<String, LoginConfigItem>();

    loginConfig['jwt'] = LoginConfigItem(
      verifier: "test2-phone-firebase-dev",
      typeOfLogin:TypeOfLogin.jwt,
      name: "test",
      clientId: 'my-ios-firebase-app-client-id',
    );


    await Web3AuthFlutter.init(Web3AuthOptions(
        clientId:
            'web3auth-plunplay-client-id',
        network:Network.testnet,
        loginConfig: loginConfig
        ));
  }

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }
```

getting the id_tokent part:

```
         final PhoneAuthCredential phoneCredential = PhoneAuthProvider.credential(
          verificationId: _verificationId!,
          smsCode: textEditingController.text,
        );

         final User? phoneUser =
             (await FirebaseAuth.instance.signInWithCredential(phoneCredential)).user;
        final idToken = await phoneUser!.getIdToken();

        final Web3AuthResponse response = await Web3AuthFlutter.login(
          LoginParams(
            loginProvider: Provider.jwt,
            extraLoginOptions: ExtraLoginOptions(
              response_type: 'token',
              verifierIdField: "sub",
              id_token: idToken,
              domain: "my-ios-app-domain",
            ),
          ),
        );
```

**Reply by shahbaz17** [reply]
Can you pass `true` in this function?

`final idToken = await phoneUser!.getIdToken(true);`

And remove `response_type: 'token',` from ExtraLoginOptions.

Also, is there a way I can see one of your idToken?

**Reply by nisa10880** [reply]
I passed true to getIdToken(true), I still get the error

**Reply by nisa10880** [reply]
I also removed response_type: 'token'

here a id_token that is just logged:

```
eyJhbGciOiJSUzI1NiIsImtpZCI6IjNmNjcyNDYxOTk4YjJiMzMyYWQ4MTY0ZTFiM2JlN2VkYTY4NDZiMzciLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJodHRwczovL3NlY3VyZXRva2VuLmdvb2dsZS5jb20vdGltZWxhcHBzLWRldiIsImF1ZCI6InRpbWVsYXBwcy1kZXYiLCJhdXRoX3RpbWUiOjE2NjY2OTAyMTMsInVzZXJfaWQiOiJrbG1VWDdkNXF2WWY3WFZDVjIxTnFUb0RKSDAzIiwic3ViIjoia2xtVVg3ZDVxdllmN1hWQ1YyMU5xVG9ESkgwMyIsImlhdCI6MTY2NjY5MDIxMywiZXhwIjoxNjY2NjkzODEzLCJwaG9uZV9udW1iZXIiOiIrMzM3NzA3MDI1MjUiLCJmaXJlYmFzZSI6eyJpZGVudGl0aWVzIjp7InBob25lIjpbIiszMzc3MDcwMjUyNSJdfSwic2lnbl9pbl9wcm92aWRlciI6InBob25lIn19.jIXdc23pnFORkCEY3856qEhDHjNcl_ijGU46vt72GiWxd5lNGh0Onzd8FoQRm6nfbKOIzGoTfW-cM9JYEec3OIhFdAqPebrNAmSghciws-XDCa283icugB4QXLc_xD10-UiHe_FlWcTIRZNOkIaJ6g1iyUiaTU9WakK_hxi9gWYIiWTCqBpj9QEOgnj1sqeI5HYNErnwqEo0jGLBGixLg9t_h5C0E-FQWQX-BZstBNgcRIz3lfFncN2LYoUMR3PXk1Df37fN-RunOKsQYBvpQnhN-G_7rH4Fijw6eTtKDJ6FgdQNtIldIqSNGCxyVPCbDh_SUFiSX0O9vyQ2-K1UZQ
```

**Reply by shahbaz17** [reply]
Can you share one of your idToken here?

**Reply by shahbaz17** [reply]
What's the JWKS endpoint for your `test2-phone-firebase-dev` verifier?

**Reply by nisa10880** [reply]
it's
https://www.googleapis.com/service_accounts/v1/jwk/securetoken@system.gserviceaccount.com

**Reply by shahbaz17** [reply]
And `JWT Validation` fields? And `JWT Verifier ID`?

**Reply by nisa10880** [reply]
you mean aud and iss values ?

**Reply by shahbaz17** [reply]
Yes, and the `JWT Verifier ID`

**Reply by nisa10880** [reply]
veriferid  : test2-phone-firebase-dev

iss : https://securetoken.google.com/com.timelapps.timelapps.dev
aud: com.timelapps.timelapps.dev

ps : I updated the id_token from previous comment

**Reply by nisa10880** [reply]
I just realized that I put the app domain name instead of firebase name 😔.  I will try with a new verifier

**Reply by shahbaz17** [reply]
Your `iss` and `aud` should be these:

```json
{
   "iss": "https://securetoken.google.com/timelapps-stg",
   "aud": "timelapps-stg"
}
```

**Reply by shahbaz17** [reply]
Yes, please try with a new verifier. Please follow https://web3auth.io/docs/custom-authentication/verifiers#create-verifier

**Reply by shahbaz17** [reply]
Please check out https://github.com/shahbaz17/web3auth-flutter-demo to see how it's done with [signInWithEmailAndPassword](https://github.com/shahbaz17/web3auth-flutter-demo/blob/master/lib/main.dart#L233)

**Reply by shahbaz17** [reply]
Any update @nisa10880 ?

**Reply by nisa10880** [reply]
Thank you very much now it's working on IOS simulator !  Version 1.03 and changing the verifier parameters fixed the issue. But when I run the app on a real device, I have an error. The web3auth webview crash (it disappears) and have this error message.

```
[AXRuntimeCommon] AX Lookup problem - errorCode:1100 error:Permission denied portName:'com.apple.iphone.axserver' PID:30497 (
	0   AXRuntime                           0x00000001da087a8c CDC17371-8478-3982-936D-2AA23E6C61DC + 326284
	1   AXRuntime                           0x00000001da03cd5c _AXGetPortFromCache + 704
	2   AXRuntime                           0x00000001da03e688 AXUIElementPerformFencedActionWithValue + 564
	3   UIKit                               0x0000000225121240 1E08AD1F-8154-3B7B-98BE-C064E7670933 + 934464
	4   libdispatch.dylib                   0x00000001b3896e6c E3EA4F63-5D11-342A-AF19-9F58DBC8E259 + 7788
	5   libdispatch.dylib                   0x00000001b3898a30 E3EA4F63-5D11-342A-AF19-9F58DBC8E259 + 14896
	6   libdispatch.dylib                   0x00000001b38a0124 E3EA4F63-5D11-342A-AF19-9F58DBC8E259 + 45348
	7   libdispatch.dylib                   0x00000001b38a0c80 E3EA4F63-5D11-342A-AF19-9F58DBC8E259 + 48256
	8   libdispatch.dylib                   0x00000001b38ab500 E3EA4F63-5D11-342A-AF19-9F58DBC8E259 + 91392
	9   libsystem_pthread.dylib             0x0000000224df10bc _pthread_wqthread + 288
	10  libsystem_pthread.dylib             0x0000000224df0e5c start_wqthread + 8
)
```

**Reply by nisa10880** [reply]
After flutter clean and pod install -repo-update and updating Xcode it is working in real IOS device.

**Comment by iNima** [comment]
Hi @shahbaz17,
Thanks for the reply.

These are the changes I applied so far:
-  I set the true parameter in the `getIdToken(true)`
- I removed `response_type: 'token'` from the ExtraLoginOptions

These are my verifier parameters:
- `Network` -> `cyan`
- `iss` -> `https://securetoken.google.com/<firebase_project_name>`
- `aud` -> `<firebase_project_name>`
- `JWT Verifier ID` -> `sub`
- `JWK Endpoint` -> `https://www.googleapis.com/service_accounts/v1/jwk/securetoken@system.gserviceaccount.com`

After running the app again I still receive this error:
`Unable to detect login share from the Auth Network. This may be due to slow internet connection. Check your internet speed and try again.`

**Reply by chaitanyapotti** [reply]
Please share the logs from your browser.
Use `chrome://inspect` to debug and share the logs here

**Comment by iNima** [comment]
I checked out [Flutter x JWT (Firebase) Example](https://github.com/Web3Auth/examples/tree/main/flutter/custom-authentication-example/using-jwt-firebase) and it worked on my device, so I knew the issue was related to my Web3Auth dashboard configuration.
I created a new custom auth verifier from scratch and it worked.
Unfortunately, on the dashboard we can't see the verifier parameters so I can't tell what parameter wasn't set correctly in the previous faulty verifier.

Thanks for you help.

**Comment by seakkus** [comment]
@iNima could you please let me know what you entered for your custom auth verifier settings - I am following the instructions here https://web3auth.io/docs/guides/firebase and getting the same error as you 'Unable to detect login share from the Auth Network. This may be due to slow internet connection. Check your internet speed and try again.' Although the user count gets increased in the dashboard, so there is something going on for sure.

Any help on this would be appreciated!

**Reply by seakkus** [reply]
I am using testnet as environment and all of the following values: 

Network -> testnet
iss -> https://securetoken.google.com/<firebase_project_id>
aud -> <firebase_project_id>
JWT Verifier ID -> sub
JWK Endpoint -> https://www.googleapis.com/service_accounts/v1/jwk/securetoken@system.gserviceaccount.com

**Reply by seakkus** [reply]
In the logs I see: 404 'https://broadcast-server.tor.us/store/get?key=0410…84a8a28c688c3f9a830635d9590e5fac45f218e5c2f9a6523' - session likely expired.

I am doing getIdToken(true) to force refresh the token.

cc: @shahbaz17

**Reply by seakkus** [reply]
It turns out to be a user error - I was passing the wrong firebase client id 🤦🏻‍♂️
